plutus-core/exe/Main.hs

```diff
@@ -218,95 +248,71 @@ exampleSingle = ExampleSingle <$> exampleName
+
 
 ---------------- Reading programs from files ----------------
 
--- Read a PLC source file
+-- Read a PLC source program
 getPlcInput :: Input -> IO String
 getPlcInput (FileInput file) = readFile file
 getPlcInput StdInput         = getContents
 
--- Read and parse a PLC source file
-parsePlcFile :: Input -> IO ParsedProgram
-parsePlcFile inp = do
-    contents <- getPlcInput inp
-    let bsContents = (BSL.fromStrict . encodeUtf8 . T.pack) contents
-    case PLC.runQuoteT $ runExceptT (PLC.parseScoped bsContents) of
-        Left (errCheck :: PlcParserError) -> do
-            T.putStrLn $ PLC.displayPlcDef errCheck
-            exitFailure
-        Right (Left (errEval :: PlcParserError)) -> do
-            print errEval
-            exitFailure
-        Right (Right (p :: ParsedProgram)) ->
-            return p
-
--- Read a PLC AST from a CBOR file
+-- Read and parse a PLC source program
+parsePlcInput :: Typing -> Input -> IO (Program PLC.AlexPosn)
+parsePlcInput typing inp = do
+    bsContents <- BSL.fromStrict . encodeUtf8 . T.pack <$> getPlcInput inp
+    case typing of
+      Typed   -> handleResult TypedProgram   $ PLC.runQuoteT $ runExceptT (PLC.parseScoped bsContents)
+      Untyped -> handleResult UntypedProgram $ PLC.runQuoteT $ runExceptT (UPLC.parseScoped bsContents)
+      where handleResult wrapper =
+                \case
+                  Left errCheck        -> failWith errCheck
+                  Right (Left errEval) -> failWith errEval
+                  Right (Right p)      -> return $ wrapper p
+            failWith (err :: PlcParserError) =  T.putStrLn (PLC.displayPlcDef err) >> exitFailure
+
+
+-- Read a CBOR-encoded PLC AST
 getCborInput :: Input -> IO BSL.ByteString
 getCborInput StdInput         = BSL.getContents
 getCborInput (FileInput file) = BSL.readFile file
 
+-- Read and deserialise a CBOR-encoded AST
+loadASTfromCBOR :: Typing -> Input -> IO (Program ())
+loadASTfromCBOR typing inp =
+    case typing of
+         Typed   -> getCborInput inp <&> deserialiseOrFail >>= handleResult TypedProgram
+         Untyped -> getCborInput inp <&> deserialiseOrFail >>= handleResult UntypedProgram
+    where handleResult wrapper =
+              \case
+                Left (DeserialiseFailure offset msg) ->
+                  putStrLn ("Deserialisation failure at offset " ++ show offset ++ ": " ++ msg) >> exitFailure
+                Right r -> return $ wrapper r
 
--- Load a PLC AST from a CBOR file
-loadPlcFromCborFile :: Input -> IO PlainProgram
-loadPlcFromCborFile inp = do
-  p <- getCborInput inp -- The type is constrained in the Right case below.
-  case deserialiseOrFail p of
-    Left (DeserialiseFailure offset msg) ->
-        do
-          putStrLn $ "Deserialisation failure at offset " ++ show offset ++ ": " ++ msg
-          exitFailure
-    Right (r::PlainProgram) -> return r
-
+-- FIXME: should we be writing error messages to stdout or stderr?
```

Comment:
Hmm, my instinct is that stderr with a nonzero exit code is the right thing to do - nobody will want to process the output error message (I think?). Not a big deal, though.

Reply:
I've done that now.  I've had to write `hPutStrLn stderr "Some error message" >> exitFailure` everywhere; I tried to abstract that into a function but it would have required type annotations at most of the call sites because there's a lot of polymorphic stuff floating about.